### DISCUSSION TITLE
How do you get Tempo changes?

### DISCUSSION THREAD

**Discussion by dreammixgames** [melanchall/drywetmidi#146] [Q&A] [answered]
Edit: I ended up realizing this was my fault after accidentally removing tempo changes when editing the midi file.

Hey! I'm working on a new original Guitar Hero inspired game called "Guitar Classic". I'm wondering how to get the changes of tempo so that I can set the fretboard speed. I'm playing GH2 songs in my game and realizing the tempo changes multiple times throughout a song.

I've tried "TempoMap.GetTempoChanges()" and that seems to get nothing, so I'm sure I just don't understand how it works.

https://youtu.be/uqgcJfh9Bk4

**Comment by melanchall** [comment, marked as answer]
Hi @dreammixgames,

If `TempoMap.GetTempoChanges()` returns nothing then there are no tempo changes. Please provide your MIDI file and I can say more.

In MIDI file times of MIDI events are written as deltas (offsets from previous events) measured in ticks. I suppose in your file those deltas become just shorter from some point in time. But that's not a tempo change. I mean tempo change is a new Set Tempo event in a file. And if `GetTempoChanges` returns nothing, there are no such events.

So I'm waiting for the file.

**Reply by dreammixgames** [reply]
Hey, it was my bad, I ended up removing the tempo changes when editing the midi file

**Reply by dreammixgames** [reply]
> Hi @dreammixgames,
> 
> If `TempoMap.GetTempoChanges()` returns nothing then there are no tempo changes. Please provide your MIDI file and I can say more.
> 
> In MIDI file times of MIDI events are written as deltas (offsets from previous events) measured in ticks. I suppose in your file those deltas become just shorter from some point in time. But that's not a tempo change. I mean tempo change is a new Set Tempo event in a file. And if `GetTempoChanges` returns nothing, there are no such events.
> 
> So I'm waiting for the file.

Hey! I figured it all out! I was using a file that I edited to remove all notes except expert, but now it's working great. Thanks for responding :)

If you're interested in seeing what I'm making, here it is: https://youtu.be/Lpe-VoP9pck

**Comment by delphiguy** [comment]
MidiFile file = MidiFile.Read("My MIDI file.mid");

Maxim, a beginner C# question on a similar topic: How, please, do I extract the ticks per quarter note from the "file" variable above? Sure, I see that that value very much exists, when I examine "file" with the debugger, but can't figure out the exact syntax to extract the value TicksPerQuarterNote from TimeDivision and assign it to (presumably) a "short" variable of my choosing--although I assume this is trivial and requires only one line of code. I've tried a million different solutions, none of which worked. Thanks for any guidance.

**Reply by melanchall** [reply]
```csharp
var ticksPerQuarterNote = ((TicksPerQuarterNoteTimeDivision)file.TimeDivision).TicksPerQuarterNote;
```

**Reply by delphiguy** [reply]
Wow.  Thanks very much.  No wonder I couldn't grasp it: I'll have to try to understand casting and inheritance, a little better.
